modules/system-view/src/main/java/org/apache/ignite/internal/systemview/SystemViewManagerImpl.java

```diff
@@ -0,0 +1,163 @@
+import static org.apache.ignite.lang.IgniteStringFormatter.format;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.ignite.internal.catalog.CatalogCommand;
+import org.apache.ignite.internal.catalog.CatalogManager;
+import org.apache.ignite.internal.catalog.commands.ColumnParams;
+import org.apache.ignite.internal.catalog.commands.CreateSystemViewCommand;
+import org.apache.ignite.internal.cluster.management.NodeAttributesProvider;
+import org.apache.ignite.internal.schema.NativeTypeSpec;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+
+/**
+ * SQL system views manager implementation.
+ */
+public class SystemViewManagerImpl implements SystemViewManager, NodeAttributesProvider {
+    public static final String NODE_ATTRIBUTES_KEY = "sql-system-views";
+
+    public static final String NODE_ATTRIBUTES_LIST_SEPARATOR = ",";
+
+    private final CatalogManager catalogManager;
+
+    private final Map<String, String> nodeAttributes = new HashMap<>();
+
+    /** Collection of system views provided by components. */
+    private final Map<String, SystemView<?>> views = new LinkedHashMap<>();
+
+    /** Busy lock to stop synchronously. */
+    private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
+
+    /** Prevents attempts to register a new system view after the component has started. */
+    private final AtomicBoolean startGuard = new AtomicBoolean();
+
+    /** Prevents double stopping of the component. */
+    private final AtomicBoolean stopGuard = new AtomicBoolean();
+
+    /** Future which is completed when system views are registered in the catalog. */
+    private final CompletableFuture<Void> schemaChangesFuture = new CompletableFuture<>();
+
+    /** Creates a system view manager. */
+    public SystemViewManagerImpl(CatalogManager catalogManager) {
+        this.catalogManager = catalogManager;
+    }
+
+    @Override
+    public void start() {
+        inBusyLock(busyLock, () -> {
+            if (!startGuard.compareAndSet(false, true)) {
+                throw new IllegalStateException("System view manager cannot be started twice");
+            }
+
+            if (views.isEmpty()) {
+                schemaChangesFuture.complete(null);
+
+                return;
+            }
+
+            catalogManager.execute(prepareCatalogCommands(views.values())).whenComplete(
+                    (r, t) -> schemaChangesFuture.complete(null)
+            );
+
+            nodeAttributes.put(NODE_ATTRIBUTES_KEY, String.join(NODE_ATTRIBUTES_LIST_SEPARATOR, views.keySet()));
+        });
+    }
+
+    @Override
+    public void stop() throws Exception {
+        if (!stopGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
```

Comment:
Should we try to complete `schemaChangesFuture` future exceptionally with e.g. `NodeStopException`?

Reply:
Definitely should, thanks.